### PR TITLE
FrequencyBandImageFilter accepts lambdas for filtering frequencies

### PR DESCRIPTION
**Pull request by dzenanz**
This filter should enable custom filtering via the functor, e.g. implementing _Butterworth_ algorithm.

### CODE REVIEW

File: Modules/Filtering/ImageFrequency/test/itkFrequencyBandImageFilterTest.cxx [hunk 1]
```diff
@@ -213,6 +213,19 @@ itkFrequencyBandImageFilterTest( int argc, char* argv[] )
   passBandShiftedFilter->SetPassBand( passLowFreqThreshold, passHighFreqThreshold );
   TRY_EXPECT_NO_EXCEPTION( passBandShiftedFilter->Update() );
 
+  // Custom functor filtering
+  auto customFunctorFilter = BandFilterType::New();
+  customFunctorFilter->SetInput( image );
+
+  auto inPlaceFunctor = []( BandFilterType::FrequencyIteratorType& freq )
```
Comment (phcerdan) [inline Modules/Filtering/ImageFrequency/test/itkFrequencyBandImageFilterTest.cxx]: can we rename `freq` to `freqIt`?

```suggestion
  auto inPlaceFunctor = []( BandFilterType::FrequencyIteratorType& freqIt )
```

File: Modules/Filtering/ImageFrequency/test/itkFrequencyBandImageFilterTest.cxx [hunk 2]
```diff
@@ -213,6 +213,19 @@ itkFrequencyBandImageFilterTest( int argc, char* argv[] )
...
+  {
+    freq.Value() *= std::sin( freq.GetFrequencyModuloSquare() * 10 );
```
Comment (phcerdan) [inline Modules/Filtering/ImageFrequency/test/itkFrequencyBandImageFilterTest.cxx]:
```suggestion
    freqIt.Value() *= std::sin( freqIt.GetFrequencyModuloSquare() * 10 );
```

**Comment by blowekamp** [comment]
Should FrequencyBandImageFilter be derived from the UnaryGeneratorFilter?

**Comment by dzenanz** [comment]
@phcerdan Do not confuse [UnaryGenerator](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/ImageFilterBase/include/itkUnaryGeneratorImageFilter.h) filter which is pretty new, and [UnaryFunctor](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkUnaryFunctorImageFilter.h) filter which is very old.

**Comment by dzenanz** [comment]
@phcerdan If backwards compatibility is not a major issue, we could refactor the filter as you propose. What do others think? @thewtex @hjmjohnson @blowekamp

**Comment by blowekamp** [comment]
+1 for separating the FrequencyBandImageFilter into a "Frequency Generator Filter" base class and a "Box Functor". They could come together into the current class similar to the "AddImageFilter" combines the BinaryGeneratorImageFilter and the AddFunctor.

**Comment by phcerdan** [comment]
I think the only module using it (heaviliy) is ITKIsotropicWavelets, and I can fix that when needed. Also, to avoid breakage of backwards compatibility, we can create the functor, and then create a FrequencyBandImageFilter just plugging together the base class and the new `boxFunctor`.

**Comment by phcerdan** [comment]
Looks like I agree with @blowekamp :smile:

**Comment by thewtex** [comment]
I don't think the frequency classes have been in a final release, so there should not be backwards compatibility concerns. :new:

**Comment by dzenanz** [comment]
Great! Let me verify that this works as intended, then I can polish it per suggestions.

**Comment by phcerdan** [comment]
By the way, `BandPassFunctor` should be a better name.

**Comment by dzenanz** [comment]
A better name for what is now known as `BoxFunctor`, or for what is now known as `FrequencyBandImageFilter`?

**Comment by phcerdan** [comment]
A better name for `BoxFunctor`.

And maybe we can add `FrequencyBandPassImageFilter`. I like that name but could be misleading because it works as well as a `FrequencyBandBlockImageFilter` depending on options...

**Comment by hjmjohnson** [comment]
Thanks all.  This is good work.

**Comment by blowekamp** [comment]
With the new GitHub workflow hopefully it is easier to manage topic branches. Please separate the nice documentation improvement in itkUnaryGeneratorFilter into separate commit to help keep a clear history.

**Comment by dzenanz** [comment]
This is working as expected. I will refactor this patch to separate the box functor into its own class and add additional Butterworth functor.

**Comment by dzenanz** [comment]
In the end I did not add Butterworth filter here, but as its [implementation](https://github.com/KitwareMedical/ITKUltrasound/blob/14cf84bc9ab0beaf8ae9d17fc3633da743e4a0d3/include/itkButterworthBandpass1DFilterFunction.h#L40-L55) is short it can be defined in appropriately adjusted form by the user.

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.hxx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    const ImageRegionType& outputRegionForThread )
+{
+  const ImageType *inputPtr = this->GetInput();
+  ImageType *outputPtr = this->GetOutput();
+
+  // Define the portion of the input to walk for this thread
+  ImageRegionType inputRegionForThread;
+  this->CallCopyOutputRegionToInputRegion(inputRegionForThread, outputRegionForThread);
+  // copy the input pixel to the output
+  ImageAlgorithm::Copy( inputPtr, outputPtr, inputRegionForThread, outputRegionForThread );
```
Comment (blowekamp) [inline Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.hxx]: If the filter is operating `InPlace` this copy should not be needed?

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+ * Please note that \ref FrequencyFFTLayoutImageRegionIteratorWithIndex requires a full FFT,
+ * and is not compatible with the hermitian optimization.
+ *
+ * To use this filter with hermitian (halved-frequency) FFTs, use
+ * \ref FrequencyHalfHermitianFFTLayoutImageRegionIteratorWithIndex or its const version.
+ *
+ * If the output of the FFT is shifted, for example after applying \ref FFTShiftImageFilter,
+ * use \ref FrequencyShiftedFFTLayoutImageRegionIteratorWithIndex.
+ *
+ *
```
Comment (blowekamp) [inline Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h]: /sa UnaryGeneratorImageFilter

File: Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx [hunk 5]
```diff
@@ -134,102 +137,83 @@ FrequencyBandImageFilter< TImageType, TFrequencyIterator >
 template< typename TImageType, typename TFrequencyIterator >
 void
 FrequencyBandImageFilter< TImageType, TFrequencyIterator >
-::DynamicThreadedGenerateData(
-  const ImageRegionType & outputRegionForThread)
+::BoxFunctor( FrequencyIteratorType& freqIt )
```
[the remark this answers is not in the document]
Comment (dzenanz) [reply Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx]: Something a bit more descriptive than just bandPass would be better, e.g. SharpBandPass, BoxBandPass, StepBandPass or something similar.

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.hxx (hunk 3 above)
Comment (dzenanz) [reply Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.hxx]: Thanks, I did not pay attention to that.

Comment (blowekamp) [reply Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.hxx]: Just review the InPlaceImageFilter, it is the `GetRunningInPlace()` method that should be called to determine if the filter is actually running inplace or not.

File: Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx (hunk 5 above)
[the remark this answers is not in the document]
Comment (phcerdan) [reply Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx]: The implementation allows two options, a band pass controlled by the modulus of the frequency value (m_RadialBand: true), or taking the absolute frequency of the pixel.

But in both cases the behaviour of the functor is to leave pixels unmodified, or set them to zero, depending on frequency. That's a PassBand, if you have a more accurate name, great to hear it, but I cannot think in a better one.

```cpp
  if ( this->m_RadialBand )
    {
    {
    if ( this->m_RadialBand )
    f = sqrt(freqIt.GetFrequencyModuloSquare());
    }
  else // Cut-off box taking into account max absolute frequency.
    {
    w = freqIt.GetFrequency();
    wMax = std::abs( *std::max_element(w.Begin(), w.End()) );
    wMin = std::abs( *std::min_element(w.Begin(), w.End()) );
    f = std::max(wMax, wMin);
    if ( wMin < wMax )
      {
      {
      f = sqrt(freqIt.GetFrequencyModuloSquare());
      freqIsNegative = true;
      }
      }
    else // Cut-off box taking into account max absolute frequency.
    }
```

[the remark this answers is not in the document]
Comment (phcerdan) [reply Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx]: Other option is to split this: if/else in different/cleaner functors, with better names.

**Comment by dzenanz** [comment]
I think this is ready for merging. Can you re-review please?

**Comment by blowekamp** [comment]
Looks good!

Does this class of filters support streaming? If not then `EnlargeOutputRequestedRegion` should be overloaded to make the output the largest possible.

**Comment by dzenanz** [comment]
@blowekamp why should the OutputRequestedRegion be enlarged?

**Comment by blowekamp** [comment]
The default behavior of the ITK pipeline (e.g. ImageToImageFilter) is to have the requested region of the output be copied to the inputs. This enables streaming for basic filters like the unary functors etc.. This filter operates with frequency space, I believe an output pixel needs more information than just the corresponding input pixel.

One way to disable streaming is to overload `EnlargeOutputRequestedRegion` to set the produced region to always be the largest possible region. By default the output region gets copied to the input. This is done when the filter only works on the whole image.

Other options depend on if and how the filter can stream.

**Comment by dzenanz** [comment]
This filter is templated over the frequency iterator. The iterator uses the meta-information to know the frequency of the currently iterated pixel. And a reference to this iterator is passed to the functor, so it can access the frequency etc. Which means that this filter should probably be able to stream. I haven't tested that, though.

**Comment by blowekamp** [comment]
> This filter is templated over the frequency iterator. The iterator uses the meta-information to know the frequency of the currently iterated pixel. And a reference to this iterator is passed to the functor, so it can access the frequency etc. Which means that this filter should probably be able to stream. I haven't tested that, though.

If the filter only has part of the input, then the boundary conditions would likely cause difference frequencies to be processes by the iterator.

It may be that the filter can produce an arbitrary output but needs the full input. In that case the `GenerateInputRequestedRegion` should be overloaded to set the input's requested region to the largest possible.

**Comment by dzenanz** [comment]
This is essentially a per-pixel filter. The only complication is that the iterator "knows" the frequency/index of the pixel, which is used as a criteria for modifying the value. I don't think the input's region needs to be expanded.

**Comment by blowekamp** [comment]
Ok, I got how the iterator operate now. Thanks!

**Comment by phcerdan** [comment]
For me, it helps to think that the input and output images are already in the frequency domain. The only role of the frequency iterators is to avoid the change of the algorithms depending on the Fourier transform or frequency layout.
So the developer thinks directly in frequencies, instead of what indices correspond to what frequencies.

**Comment by dzenanz** [comment]
@thewtex I fixed that with the latest push. All checks are now passing.

File: Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx [hunk 6]
```diff
@@ -134,102 +137,83 @@ FrequencyBandImageFilter< TImageType, TFrequencyIterator >
...
   typename FrequencyIteratorType::FrequencyType w;
   FrequencyValueType wMax;
   FrequencyValueType wMin;
   bool freqIsNegative = false;
 
-  while ( !freqIt.IsAtEnd() )
+  if ( this->m_RadialBand )
     {
-    if ( this->m_RadialBand )
+    f = sqrt(freqIt.GetFrequencyModuloSquare());
```
Comment (thewtex) [inline Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx]: `f" -> `frequency`

Please use descriptive names, non-single character names, that follow the ITK style guidelines for readability and so the codebase is searchable.

File: Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx [hunk 7]
```diff
@@ -134,102 +137,83 @@ FrequencyBandImageFilter< TImageType, TFrequencyIterator >
...
+    }
+  else // Cut-off box taking into account max absolute frequency.
+    {
+    w = freqIt.GetFrequency();
```
Comment (thewtex) [inline Modules/Filtering/ImageFrequency/include/itkFrequencyBandImageFilter.hxx]: `w`?

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+ * A frequency filtering functor needs to be supplied via one of SetFunctor() overloads.
+ * The functor should take FrequencyIteratorType reference as its only parameter.
+ * If functor configurability is required,
+ * those parameters should be passed directly to the functor at the place of definition.
+ *
+ *
+ * Filters in the module ITKImageFrequency work with input images in the frequency domain.
+ * This filter is templated over a TFrequencyIterator depending on the
+ * frequency layout of the input image.
+ * Images in the dual space can be acquired experimentally, from scattering exaperiments or other techniques.
```
Comment (thewtex) [inline Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h]: expaperiments -> experiments

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+ * The layout of these images is the same as spatial domain images. Use \ref FrequencyImageRegionIteratorWithIndex
+ *
+ * Frequency-domain images can be computed from any spatial-domain applying a Fourier Transform \ref ForwardFFTImageFilter.
+ * Use \ref FrequencyFFTLayoutImageRegionIteratorWithIndex.
+ * Please note that \ref FrequencyFFTLayoutImageRegionIteratorWithIndex requires a full FFT,
+ * and is not compatible with the hermitian optimization.
```
Comment (thewtex) [inline Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h]: hermitian -> Hermitian

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+ * The functor should take FrequencyIteratorType reference as its only parameter.
+ * If functor configurability is required,
+ * those parameters should be passed directly to the functor at the place of definition.
+ *
+ *
+ * Filters in the module ITKImageFrequency work with input images in the frequency domain.
+ * This filter is templated over a TFrequencyIterator depending on the
+ * frequency layout of the input image.
+ * Images in the dual space can be acquired experimentally, from scattering exaperiments or other techniques.
+ * The layout of these images is the same as spatial domain images. Use \ref FrequencyImageRegionIteratorWithIndex
```
Comment (thewtex) [inline Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h]: What is the meaning of this sentence? Can it be reworded to clarify that meaning?

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+ *
+ * Frequency-domain images can be computed from any spatial-domain applying a Fourier Transform \ref ForwardFFTImageFilter.
+ * Use \ref FrequencyFFTLayoutImageRegionIteratorWithIndex.
```
Comment (thewtex) [inline Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h]: Use -> If \ref ForwardFFTImageFilter was used for the domain transformation, template this filter with the \ref FrequencyFFTLayoutImageRegionIteratorWithIndex.

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h (hunk 10 above)
Comment (dzenanz) [reply Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h]: I reworded it. @phcerdan could double-check.

Comment (thewtex) [reply Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h]: Thanks!

File: Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h (hunk 11 above)
Comment (thewtex) [reply Modules/Filtering/ImageFrequency/include/itkUnaryFrequencyDomainFilter.h]: > Note: Should we then enumerate here all the possibilities? FFTHalfHermitian, FFTShifted?

:+1: good idea.

**Comment by phcerdan** [comment]
Sorry, I was unplugged on the weekend, sorry for not answering on time. But I would have approved,  `scalarFrequency` is a way better name than the old `w`. Thanks @dzenanz

**Pull request merged**